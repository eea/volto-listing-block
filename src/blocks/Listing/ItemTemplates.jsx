import cx from 'classnames';
import { ConditionalLink, FormattedDate } from '@plone/volto/components';

import PreviewImage from '@eeacms/volto-listing-block/PreviewImage';

const BodyText = ({ item, hasDate, hasDescription }) => {
  return (
    <div className="listing-body">
      <h3 className={'listing-header'}>{item.title ? item.title : item.id}</h3>
      {hasDate && item.effective && (
        <p className={'listing-date'}>
          <FormattedDate
            date={item.effective}
            format={{
              year: 'numeric',
              month: 'short',
              day: 'numeric',
            }}
          />
        </p>
      )}
      {hasDescription && (
        <p className={'listing-description'}>{item.description}</p>
      )}
    </div>
  );
};

const getStyles = (props) => {
  const { itemModel = {} } = props;
  const res = {};
  if (itemModel.maxDescription) {
    res[`max-${itemModel.maxDescription}-lines`] = true;
  }
  return res;
};

const BasicItem = (props) => {
  const { item, styles, className, itemModel = {}, isEditMode = false } = props;
  const { hasImage, imageOnRightSide, hasDate, hasDescription } = itemModel;

  return (
    <div
      className={cx(
        'u-item listing-item',
        styles?.theme,
        getStyles(props),
        className,
      )}
    >
      <ConditionalLink item={item} condition={!isEditMode}>
        {hasImage ? (
          imageOnRightSide ? (
            <>
              <BodyText
                item={item}
                hasDescription={hasDescription}
                hasDate={hasDate}
              />
              <PreviewImage item={item} style={{ marginLeft: 'auto' }} />
            </>
          ) : (
            <>
              <PreviewImage item={item} />
              <BodyText
                item={item}
                hasDescription={hasDescription}
                hasDate={hasDate}
              />
            </>
          )
        ) : (
          <BodyText
            item={item}
            hasDescription={hasDescription}
            hasDate={hasDate}
          />
        )}
      </ConditionalLink>
    </div>
  );
};

export const DefaultItemLayout = (props) => {
  return <BasicItem {...props} />;
};
