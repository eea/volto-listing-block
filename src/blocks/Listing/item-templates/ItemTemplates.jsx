import cx from 'classnames';
import { ConditionalLink, FormattedDate } from '@plone/volto/components';
// import { buildStyleClassNamesFromData } from '@plone/volto/helpers';

import { getVoltoStyles } from '@eeacms/volto-listing-block/schema-utils';

import PreviewImage from '@eeacms/volto-listing-block/PreviewImage';

const getStyles = (props) => {
  const { itemModel = {} } = props;
  const res = {};
  if (itemModel.maxDescription) {
    res[`max-${itemModel.maxDescription}-lines`] = true;
  }
  if (itemModel.maxTitle) {
    res[`title-max-${itemModel.maxTitle}-lines`] = true;
  }
  return res;
};

const BodyText = ({ item, hasDate, hasDescription }) => {
  return (
    <div className="listing-body">
      <h3 className={'listing-header'}>{item.title ? item.title : item.id}</h3>
      {hasDate && item.effective && (
        <p className={'listing-date'}>
          <FormattedDate
            date={item.effective}
            format={{
              year: 'numeric',
              month: 'short',
              day: 'numeric',
            }}
          />
        </p>
      )}
      {hasDescription && (
        <p className={'listing-description'}>{item.description}</p>
      )}
    </div>
  );
};

const BasicItem = (props) => {
  const { item, className, itemModel = {}, isEditMode = false } = props;
  const { hasImage, imageOnRightSide, hasDate, hasDescription } = itemModel;
  const styles = getStyles(props);

  return (
    <div
      className={cx('u-item listing-item', getVoltoStyles(styles), className)}
    >
      <div
        className={`wrapper ${imageOnRightSide ? 'right-image' : 'left-image'}`}
      >
        <div className="slot-top">
          <ConditionalLink item={item} condition={!isEditMode}>
            {hasImage ? (
              imageOnRightSide ? (
                <>
                  <BodyText
                    item={item}
                    hasDescription={hasDescription}
                    hasDate={hasDate}
                  />
                  <div className="image-wrapper">
                    <PreviewImage item={item} />
                  </div>
                </>
              ) : (
                <>
                  <div className="image-wrapper">
                    <PreviewImage item={item} />
                  </div>
                  <BodyText
                    item={item}
                    hasDescription={hasDescription}
                    hasDate={hasDate}
                  />
                </>
              )
            ) : (
              <BodyText
                item={item}
                hasDescription={hasDescription}
                hasDate={hasDate}
              />
            )}
          </ConditionalLink>
        </div>
        <div className="slot-bottom">{item?.extra}</div>
      </div>
    </div>
  );
};

export const DefaultItemLayout = (props) => {
  return <BasicItem {...props} />;
};
