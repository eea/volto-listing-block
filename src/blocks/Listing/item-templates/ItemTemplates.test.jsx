import React from 'react';
import { render, screen } from '@testing-library/react';
import { DefaultItemLayout } from './ItemTemplates';
import '@testing-library/jest-dom/extend-expect';

jest.mock('@plone/volto/components', () => ({
  ConditionalLink: () => <div>ConditionalLink</div>,
  UniversalLink: () => <div>UniversalLink:</div>,
}));

describe('ItemTemplates', () => {
  const item = {
    title: 'Default listing title',
    description: 'Default listing description',
    meta: 'Default listing meta',
    EffectiveDate: '2023-10-05T08:21:00+02:00',
    url: '/my-item-url',
  };
  it('renders correctly', () => {
    const itemModel = {
      hasImage: true,
      hasDate: true,
      hasDescription: true,
      imageOnRightSide: false,
    };

    const { container } = render(
      <DefaultItemLayout
        item={item}
        itemModel={itemModel}
        className="my-class"
      />,
    );

    const itemContainer = container.querySelector(
      '.u-item.listing-item.my-class',
    );
    expect(itemContainer).toBeInTheDocument();

    const image = container.querySelector('img');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('alt', 'Default listing title');
    expect(image).toHaveClass('ui image');

    expect(screen.getByText('ConditionalLink')).toBeInTheDocument();

    expect(screen.getByText('05 Oct 2023')).toHaveClass('listing-date');

    expect(screen.getByText('Default listing description')).toHaveClass(
      'listing-description',
    );
  });

  it('renders correctly without a description', () => {
    const itemModel = { hasImage: true, hasDate: true, hasDescription: false };

    render(<DefaultItemLayout item={item} itemModel={itemModel} />);

    expect(
      screen.queryByText('Default listing description'),
    ).not.toBeInTheDocument();
  });

  it('renders without an image if hasImage is false', () => {
    const itemModel = { hasImage: false, hasDate: true, hasDescription: true };

    const { container } = render(
      <DefaultItemLayout item={item} itemModel={itemModel} />,
    );

    expect(container.querySelector('.image-wrapper')).not.toBeInTheDocument();
  });

  it('renders with the image on the right when imageOnRightSide is true', () => {
    const itemModel = {
      hasImage: true,
      hasDate: true,
      hasDescription: true,
      imageOnRightSide: true,
    };

    const { container } = render(
      <DefaultItemLayout item={item} itemModel={itemModel} />,
    );

    const wrapper = container.querySelector('.wrapper');
    expect(wrapper).toHaveClass('right-image');

    const slotTop = container.querySelector('.slot-top');
    const bodyText = slotTop.querySelector('.listing-body');
    const imageWrapper = slotTop.querySelector('.image-wrapper');

    expect(bodyText).toBe(slotTop.firstChild);
    expect(imageWrapper).toBe(slotTop.lastChild);
  });
});
