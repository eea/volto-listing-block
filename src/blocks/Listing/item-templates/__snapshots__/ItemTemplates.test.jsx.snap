// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemTemplates renders correctly 1`] = `
<div
  className="u-item listing-item my-class"
>
  <div
    className="wrapper left-image"
  >
    <div
      className="slot-top"
    >
      <div
        className="image-wrapper"
      >
        <img
          alt="Default listing title"
          className="ui image"
          src="default-image.svg"
        />
      </div>
      <div
        className="listing-body"
      >
        <div>
          ConditionalLink
        </div>
        <div
          className="listing-body-dates"
        >
          <p
            className="listing-date"
          >
            05 Oct 2023
          </p>
        </div>
        <p
          className="listing-description"
        >
          Default listing description
        </p>
      </div>
    </div>
    <div
      className="slot-bottom"
    />
  </div>
</div>
`;
